### PR TITLE
Issue #198 : Support alternative URL returned by images API

### LINKED ISSUE #198: IndexOutOfBoundsException in RemoteMediaStore after failing to parse images.ala.org.au URL
There is an IndexOutOfBoundsException that can occur, based on a stacktrace and checking the code, in RemoteMediaStore.save:

```
2017-03-16 04:43:04,121 ERROR: [RemoteMediaStore] - Problem downloading media. URL:http://www2.ala.org.au/datasets/occurrence/dr341/associatedMedia/X00105.mp3
2017-03-16 04:43:04,122 WARN : [DataLoader] - Unable to save file: http://www2.ala.org.au/datasets/occurrence/dr341/associatedMedia/X00105.mp3
2017-03-16 04:43:04,122 INFO : [RemoteMediaStore] - Remote media store URL recognised: http://images.ala.org.au/store/e/7/f/3/eb024033-4da4-4124-83f7-317365783f7e/original
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index: 0
	at java.util.Collections$EmptyList.get(Collections.java:3212)
	at au.org.ala.biocache.load.RemoteMediaStore$.save(MediaStore.scala:258)
	at au.org.ala.biocache.load.DataLoader$$anonfun$processMedia$1.apply(DataLoader.scala:317)
	at au.org.ala.biocache.load.DataLoader$$anonfun$processMedia$1.apply(DataLoader.scala:296)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at au.org.ala.biocache.load.DataLoader$class.processMedia(DataLoader.scala:296)
	at au.org.ala.biocache.load.DwCALoader.processMedia(DwCALoader.scala:89)
	at au.org.ala.biocache.load.DwCALoader.loadArchive(DwCALoader.scala:295)
	at au.org.ala.biocache.load.DwCALoader.loadLocal(DwCALoader.scala:145)
	at au.org.ala.biocache.load.DwCALoader$.main(DwCALoader.scala:77)
	at au.org.ala.biocache.cmd.CMD2$.main(CMD2.scala:134)
	at au.org.ala.biocache.cmd.CMD2.main(CMD2.scala)
```

The cause seems to be related to URLEncodedUtils attempting to parse and possibly split the URL based on a presumed pattern but failing and instead returning an empty List:

```
  //  http://images.ala.org.au/image/proxyImageThumbnailLarge?imageId=119d85b5-76cb-4d1d-af30-e141706be8bf
 val uri = new URI(urlToMedia)
 val params:java.util.List[NameValuePair] = URLEncodedUtils.parse(uri, "UTF-8")
 val param = params.get(0)
```

The logger message that occurred before the stacktrace indicates that the URL which was returned from the images.ala.org.au API does not match the presumed URL structure so there needs to be new work done to parse the URLs that images.ala.org.au is returning in this case:

```
2017-03-16 04:43:04,122 INFO : [RemoteMediaStore] - Remote media store URL recognised: http://images.ala.org.au/store/e/7/f/3/eb024033-4da4-4124-83f7-317365783f7e/original
```

This is blocking the reloading/reprocessing of the ANWC Sound collection, tracked at https://github.com/AtlasOfLivingAustralia/data-management/issues/296

### PR DESCRIPTION
**Pull request by ansell**
The images API returns various URL forms containing its internal UUID identifier. This pull request attempts to support another of the variations, those containing "/store/" and the UUID in the path, rather than as a query parameter as was supported in the previous version.

This fixes #198

### CODE REVIEW

**Comment by ansell** [comment]
@adam-collins could you review these changes if you are still working on biocache-store today?

File: src/main/scala/au/org/ala/biocache/load/MediaStore.scala [hunk 1]
```diff
@@ -252,16 +253,48 @@ object RemoteMediaStore extends MediaStore {
...
+            }
+          }
+        }
+        None
+      }
+      else {
+        None
+      }
+      
+      if(imageId.isEmpty) {
```
Comment (adam-collins) [inline src/main/scala/au/org/ala/biocache/load/MediaStore.scala]: When the url starts with Config.remoteMediaStoreUrl the subsequent unchanged code, metadata update or download/upload, in this method does not apply. The previous code had a return statement that should also have returned None when imageId.isEmpty == true.

**Comment by adam-collins** [comment]
In addition to the code comment about the early 'return', the imageId is always set to None.

A test case:
```
test("save media store url id detection") {

    val tests = List(("test valid image/proxyImage", Config.remoteMediaStoreUrl + "/image/proxyImageThumbnailLarge?imageId=119d85b5-76cb-4d1d-af30-e141706be8bf",true),
      ("test invalid image/proxyImage, no imageId",Config.remoteMediaStoreUrl + "/image/proxyImageThumbnailLarge?id=119d85b5-76cb-4d1d-af30-e141706be8bf", false),
      ("test valid store" ,Config.remoteMediaStoreUrl + "/store/e/7/f/3/eb024033-4da4-4124-83f7-317365783f7e/original", true),
      ("test invalid store, bad UUID",Config.remoteMediaStoreUrl + "/store/e/7/f/3/eb024033-4da4-414-83f7-317365783f7e/original", false))

    tests.foreach { it =>
      logger.info(it._1)
      expectResult(it._3) {
        Config.mediaStore.save("", "", it._2, None) match {
          case Some((filename, filepath)) => true
          case None => false
        }
      }
    }
  }
```

**Comment by ansell** [comment]
Thanks for the test cases, it appears that UUID.fromString does not fail on invalid UUIDs like in the test case above, it attempts to patch them before parsing, so will need to investigate another way of verifying the UUID.

**Comment by ansell** [comment]
Hi @adam-collins I fixed the merge conflict on the .gitignore file and added some tests for the UUID verification. Can you review this again?

**Pull request merged**